### ISSUE TITLE
Illegal mix of collations (utf8mb4_unicode_ci,IMPLICIT) and (utf8mb4_general_ci,IMPLICIT) in 0075_populate_latest_revision_and_revision_object_str

### ISSUE THREAD

**Issue by konstin** [wagtail/wagtail#9477]
### Issue Summary

When migrating a django project (on a fresh db) that uses mariadb:10 with  `--character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci`, the migration `0075_populate_latest_revision_and_revision_object_str` of wagtailcore fails with the following traceback. I unfortunately failed to find out which two were the offending fields. I also couldn't find any db settings to make the migration pass anyway

```
File "/app/.venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 89, in _execute                                                                                             
    return self.cursor.execute(sql, params)                                                                                                                                                   
  File "/app/.venv/lib/python3.8/site-packages/django/db/backends/mysql/base.py", line 75, in execute                                                                                         
    return self.cursor.execute(query, args)                                                                                                                                                   
  File "/app/.venv/lib/python3.8/site-packages/MySQLdb/cursors.py", line 206, in execute                                                                                                      
    res = self._query(query)                                                                                                                                                                  
  File "/app/.venv/lib/python3.8/site-packages/MySQLdb/cursors.py", line 319, in _query                                                                                                       
    db.query(q)                                                                                                                                                                               
  File "/app/.venv/lib/python3.8/site-packages/MySQLdb/connections.py", line 254, in query                                                                                                    
    _mysql.connection.query(self, query)                                                                                                                                                      
MySQLdb.OperationalError: (1267, "Illegal mix of collations (utf8mb4_unicode_ci,IMPLICIT) and (utf8mb4_general_ci,IMPLICIT) for operation '='")                                               

The above exception was the direct cause of the following exception:
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./manage.py", line 27, in <module>
    main()
  File "./manage.py", line 23, in main
    execute_from_command_line(sys.argv)
  File "/app/.venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/app/.venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/app/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 402, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/app/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 448, in execute
    output = self.handle(*args, **options)
  File "/app/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 96, in wrapped
    res = handle_func(*args, **kwargs)
  File "/app/.venv/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 349, in handle
    post_migrate_state = executor.migrate(
  File "/app/.venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/app/.venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/app/.venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/app/.venv/lib/python3.8/site-packages/django/db/migrations/migration.py", line 125, in apply
    operation.database_forwards(
  File "/app/.venv/lib/python3.8/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/app/.venv/lib/python3.8/site-packages/wagtail/migrations/0075_populate_latest_revision_and_revision_object_str.py", line 18, in populate_latest_revision
    Page.objects.all().update(latest_revision_id=latest_revision_id)
  File "/app/.venv/lib/python3.8/site-packages/django/db/models/query.py", line 1190, in update
    rows = query.get_compiler(self.db).execute_sql(CURSOR)
  File "/app/.venv/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1822, in execute_sql
    cursor = super().execute_sql(result_type)
  File "/app/.venv/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1398, in execute_sql
    cursor.execute(sql, params)
  File "/app/.venv/lib/python3.8/site-packages/sentry_sdk/integrations/django/__init__.py", line 562, in execute
    return real_execute(self, sql, params)
  File "/app/.venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 67, in execute                                                                                             
    return self._execute_with_wrappers(
  File "/app/.venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers                                                                              
    return executor(sql, params, many, context)
  File "/app/.venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 89, in _execute                                                                                            
    return self.cursor.execute(sql, params)
  File "/app/.venv/lib/python3.8/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/app/.venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 89, in _execute                                                                                            
    return self.cursor.execute(sql, params)
  File "/app/.venv/lib/python3.8/site-packages/django/db/backends/mysql/base.py", line 75, in execute                                                                                        
    return self.cursor.execute(query, args)
  File "/app/.venv/lib/python3.8/site-packages/MySQLdb/cursors.py", line 206, in execute
    res = self._query(query)
  File "/app/.venv/lib/python3.8/site-packages/MySQLdb/cursors.py", line 319, in _query
    db.query(q)
  File "/app/.venv/lib/python3.8/site-packages/MySQLdb/connections.py", line 254, in query
    _mysql.connection.query(self, query)
django.db.utils.OperationalError: (1267, "Illegal mix of collations (utf8mb4_unicode_ci,IMPLICIT) and (utf8mb4_general_ci,IMPLICIT) for operation '='")                                      
```

### Steps to Reproduce

1. Start a mariadb:10 docker container with `--character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci`
2. `git clone https://github.com/meine-stadt-transparent/meine-stadt-transparent`
3. `cd meine-stadt-transparent`
4. `git checkout 7af8a4ee514c8d19fc7b54ca41ae372c8fe8eede`
5. `poetry install` # tested with poetry==1.2.2
6. `./manage.py migrate wagtailcore`

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: Unfortunately, I failed to reproduce this with a fresh project. I think it needs the previous migrations to cause the bug. It happens with a completely fresh database when only migrating wagtailcore though

### Technical details

-   Python version: Run `python --version`. `python:3.8` docker container
-   Django version: Look in your requirements.txt, or run `pip show django | grep Version`. 4.1.2 (in poetry.lock)
-   Wagtail version: Look at the bottom of the Settings menu in the Wagtail admin, or run `pip show wagtail | grep Version:`. 4.0.2 (in poetry.lock)
-   Browser version: n/a

**Comment by vsalvino** [comment]
So technically this is not a bug in Wagtail, but it is related to some very specific behavior of MySQL and in particular, `mysqlclient`.

I'm working on a PR that will add an option to `mysqlclient` to fix this behavior.

https://github.com/PyMySQL/mysqlclient/issues/563

**Comment by gasman** [comment]
@vsalvino Does this mean it would be reasonable to close this issue? Or are we looking at implementing a workaround on the Wagtail side as well as addressing it in mysqlclient?

**Comment by vsalvino** [comment]
I don't think any action is necessary in Wagtail. Wagtail is using legal Django/SQL syntax.

However it would be worth adding a note to the docs, if and when mysqlclient accepts the change. I am planning to add such a note to both Wagtail and Django docs whenever the patch is released in mysqlclient.

It may be useful to temporarily add a note to the Wagtail 4 upgrade considerations, as this behavior will affect a large portion of MySQL users.

**Comment by vanschelven** [comment]
Writing this in pure Python, using query-in-loop rather than a subquery, would be a workable workaround, right? I understand the performance would be worse, but it would at least allow users to move forward. Perhaps this workaround could be put into the `except` clause that catches this specific `OperationalError`?

**Comment by laymonage** [comment]
> The offending query involves using a `CAST`, which requires a specific character set and collation. If the character set and collation is omitted from the `CAST` statement (which is how the Django ORM generates it - rightfully so as the ORM is not aware of collations), the connection default is used. Due to a behavior of `mysqlclient`, it always sets the character set when connecting... but not the collation... which means the collation falls back to the hard-coded default in MySQL/MariaDB - _not_ the value set as the server default.

`CharField`/`TextField` have a [`db_collation`](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.CharField.db_collation) parameter, I wonder if that also works when used in a `Cast()`?

If so, maybe we could use `Cast()` to force both the left-hand-side and right-hand-side to use the same collation, if `schema_editor.connection.vendor == "mysql"`?

**Comment by vsalvino** [comment]
I don't think setting `db_collation` would help. In this case, we have two tables/columns which are both using the exact same collation. However the connection is using an entire different and incorrect collation. The CAST statement uses the connection's collation, not the table's/column's. This is a documented behavior of MySQL and MariaDB. The Django ORM does not add custom collation or charset args to the CAST statement, even if you set `db_collation` on the fields.

The only workaround is to not use a Cast statement in Django, or to use my patch to mysqlclient which correctly sets the collation of the connection. If you have the rare setup where your tables/columns are intentionally using different collations, then there would be no possible way to use the Cast statement in Django between those columns.

I have not been able to get a response from the maintainer of `mysqlclient`.

**Comment by vanschelven** [comment]
I had thought (hoped) that managing to get past the migration stage with manual effort would be a workable workaround but I'm afraid it isn't. There's also a lot of usage of this function in the live (non-migration) code. As of today:

```
$ find -name "*.py" | x grep 'from django.db.models.functions import.*Cast'
./wagtail/blocks/migrations/migrate_operation.py:from django.db.models.functions import Cast
./wagtail/tests/streamfield_migrations/test_migrations.py:from django.db.models.functions import Cast
./wagtail/admin/views/generic/models.py:from django.db.models.functions import Cast
./wagtail/models/__init__.py:from django.db.models.functions import Cast, Concat, Substr
./wagtail/migrations/0075_populate_latest_revision_and_revision_object_str.py:from django.db.models.functions import Cast
./wagtail/migrations/0071_populate_revision_content_type.py:from django.db.models.functions import Cast
./wagtail/query.py:from django.db.models.functions import Cast, Length, Substr
./wagtail/search/backends/database/postgres/postgres.py:from django.db.models.functions import Cast, Length
./wagtail/search/backends/database/sqlite/sqlite.py:from django.db.models.functions import Cast, Length
./wagtail/search/models.py:from django.db.models.functions import Cast
```

**Comment by gasman** [comment]
Just to note that the use of `Cast` in wagtail/search/models.py has been there since 2.15 - so if `Cast` in itself is a problem, then this potentially spans more than just 4.0 (and so a note in the 4.0 release notes may not be sufficient).

**Comment by vanschelven** [comment]
> Just to note that the use of Cast in wagtail/search/models.py has been there since 2.15 

True, but in https://github.com/wagtail/wagtail/pull/8441 there has been a great number of added `Cast` usages.

**Comment by benwebber** [comment]
We're using a [`connection_created`](https://docs.djangoproject.com/en/dev/ref/signals/#connection-created) signal receiver to work around this issue without downgrading or patching mysqlclient.

This executes *after* the problem `SET NAMES` statement mentioned in @vsalvino's PR.

```python
@receiver(connection_created)
def connection_created_set_collation_connection(sender, connection, **kwargs):
    if connection.vendor != 'mysql':
        return
    with connection.cursor() as cursor:
        cursor.execute('SET collation_connection = utf8mb4_unicode_ci')
```

**Comment by vsalvino** [comment]
Update: we have published a fork of mysqlclient to PyPI that fixes this problem. We are also going to start shipping this fork by default on CodeRed Cloud to prevent the creation of broken Wagtail sites ad infinitum.

https://pypi.org/project/mysqlclient-collate/

**Comment by vsalvino** [comment]
Update: the fix for this was added to `mysqlclient==2.2.0`. Please use that version, and set the appropriate `charset` and `collation` parameters.

```
pip install mysqlclient>=2.2
```

```python
# settings.py

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.mysql",
        "HOST": "",
        "NAME": "",
        "USER": "",
        "PASSWORD": "",
        "OPTIONS": {
            "charset": "utf8mb4",
            "collation": "utf8mb4_unicode_ci",
        },
    }
```

This issue can probably be closed now.
